plugins/fix-china-map-offset.js

```diff
@@ -0,0 +1,313 @@
+
+    var radLat = wgs.lat / 180 * Math.PI;
+    var magic = 1 - GCJ_EE * Math.pow(Math.sin(radLat), 2); // just a common expr
+
+    // Arc lengths per degree, on the wrong ellipsoid
+    var lat_deg_arclen = Math.PI / 180 * (GCJ_A * (1 - GCJ_EE)) * Math.pow(magic, 1.5);
+    var lon_deg_arclen = Math.PI / 180 * (GCJ_A * Math.cos(radLat) / Math.sqrt(magic));
+
+    return {
+      lat: wgs.lat + dLat_m / lat_deg_arclen,
+      lng: wgs.lng + dLon_m / lon_deg_arclen,
+    };
+  };
+
+  return gcjObfs;
+})();
+
+plugin.fixChinaOffset.process = function (wgs) {
+  return plugin.fixChinaOffset.isInGoogle(wgs)
+    ? plugin.fixChinaOffset.gcjObfs(wgs)
+    : wgs;
+};
+
+///////// begin overwrited L.GridLayer /////////
+
+L.GridLayer.prototype._getTiledPixelBounds = (function () {
+  return function (center) {
+  /// modified here ///
+    center = window.plugin.fixChinaOffset.getLatLng(center, this.options);
+    /////////////////////
+    var map = this._map,
+      mapZoom = map._animatingZoom ? Math.max(map._animateToZoom, map.getZoom()) : map.getZoom(),
+      scale = map.getZoomScale(mapZoom, this._tileZoom),
+      pixelCenter = map.project(center, this._tileZoom).floor(),
+      halfSize = map.getSize().divideBy(scale * 2);
+
+    return new L.Bounds(pixelCenter.subtract(halfSize), pixelCenter.add(halfSize));
+  };
+})(L.GridLayer.prototype._getTiledPixelBounds);
+
+L.GridLayer.prototype._setZoomTransform = (function (original) {
+  return function (level, center, zoom) {
+    center = window.plugin.fixChinaOffset.getLatLng(center, this.options);
+    original.apply(this, [level, center, zoom]);
+  };
+})(L.GridLayer.prototype._setZoomTransform);
+
+L.GridLayer.GoogleMutant.prototype._update = (function (original) {
+  return function () {
+    this.options.needFixChinaOffset = true;
+    // zoom level check needs to happen before super's implementation (tile addition/creation)
+    // otherwise tiles may be missed if maxNativeZoom is not yet correctly determined
+    if (this._mutant) {
+      var center = this._map.getCenter();
+      /// modified here ///
+      center = window.plugin.fixChinaOffset.getLatLng(center, this.options);
+      /////////////////////
+      var _center = new google.maps.LatLng(center.lat, center.lng);
+
+      this._mutant.setCenter(_center);
+      var zoom = this._map.getZoom();
+      var fractionalLevel = zoom !== Math.round(zoom);
+      var mutantZoom = this._mutant.getZoom();
+
+      //ignore fractional zoom levels
+      if (!fractionalLevel && (zoom != mutantZoom)) {
+        this._mutant.setZoom(zoom);
+
+        if (this._mutantIsReady) this._checkZoomLevels();
+        //else zoom level check will be done later by 'idle' handler
+      }
+    }
+
+    L.GridLayer.prototype._update.call(this);
+  };
+})(L.GridLayer.GoogleMutant.prototype._update);
+
+window.plugin.fixChinaOffset.getLatLng = function (pos, options) {
+  // No offsets in satellite and hybrid maps
```

Comment:
It is true for Google map, but it's not reliable to to extrapolate such assumption for arbitrary map.

So I propose to check real maps.
And may be we need additional configuration option.
```suggestion
  // No offsets in satellite and hybrid maps
  // TODO: it's true for Google maps, need to examine other available Chinese maps
```

Reply:
No, this fix is true for all cards that have a correction.

Another thing is that in exceptional cases, when map is connected not by Leaflet, but by means of its own libraries, it may be necessary to correct loading of tiles in them. Only loading, but not coordinates.

Reply:
> No, this fix is true for all cards that have a correction.

Well, I mean next line:
```
  if (options.needFixChinaOffset === true && options.type !== 'satellite' && options.type !== 'hybrid') {
```
Are you sure that in other maps we also should skip correction for satellite and hybrid?

> when map is connected not by Leaflet

I suppose that it also could be fixed. But let's not run ahead.

Reply:
https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China:
> Some map providers, such as Here, choose to also offset their satellite imagery layer to match the GCJ-02 street map.[34]

So we need more flexible options.